Add OpenTelemetry library and expose metrics for Witsml requests

lindsve opened the pull request:
## Fixes

#2290 

Implements OpenTelemetry metrics:
- witsml_requests_duration_seconds (histogram)
- witsml_requests_active (counter)

## Description

The histogram has been set up with the following buckets: 0.5, 1, 3, 5, 10, 15, 30 and 60 seconds.

When used in an application using the library, the metrics can be added within the `AddOpenTelemetry` method added by the OpenTelementry SDK:

```
services.AddOpenTelemetry()
            .WithMetrics(builder =>
            {
                builder.SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("MyApplication"));
                builder.AddWitsmlInstrumentation();

                builder.AddPrometheusExporter();
            });
```

Example from prometheus export:
```
# TYPE witsml_requests_duration_seconds histogram
# UNIT witsml_requests_duration_seconds seconds
# HELP witsml_requests_duration_seconds Time spent during requests to a Witsml server
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="0.5"} 31251 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="1"} 31691 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="3"} 31911 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="5"} 31941 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="10"} 31943 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="15"} 31944 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="30"} 31944 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="60"} 31944 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1",le="+Inf"} 31945 1710160333618
witsml_requests_duration_seconds_sum{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1"} 1171 1710160333618
witsml_requests_duration_seconds_count{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="1"} 31945 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="0.5"} 27621 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="1"} 27876 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="3"} 27979 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="5"} 27993 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="10"} 27993 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="15"} 27993 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="30"} 27994 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="60"} 27994 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1",le="+Inf"} 27994 1710160333618
witsml_requests_duration_seconds_sum{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1"} 563 1710160333618
witsml_requests_duration_seconds_count{host="witsml.somewhere.com",method="GetFromStore",objectType="risk",resultCode="1"} 27994 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="0.5"} 2 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="1"} 2 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="3"} 2 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="5"} 2 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="10"} 2 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="15"} 2 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="30"} 2 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="60"} 2 1710160333618
witsml_requests_duration_seconds_bucket{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2",le="+Inf"} 2 1710160333618
witsml_requests_duration_seconds_sum{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2"} 0 1710160333618
witsml_requests_duration_seconds_count{host="witsml.somewhere.com",method="GetFromStore",objectType="log",resultCode="2"} 2 1710160333618

# TYPE witsml_requests_active gauge
# HELP witsml_requests_active Number of active requests
witsml_requests_active{host="witsml.somewhere.com",method="GetFromStore",objectType="log"} 1 1710160333618
witsml_requests_active{host="witsml.somewhere.com",method="GetFromStore",objectType="risk"} 2 1710160333618
```

All metrics are tagged with host, method and objectType. If available, resultCode is also added as a tag.

Note: This has only been tested in a .NET 8 application.

## Type of change

* [ ] Bugfix
* [x] New feature (non-breaking change which adds functionality)
* [ ] Enhancement of existing functionality
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [ ] Frontend
* [ ] API
* [x] WITSML
* [ ] Other (please describe)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [x] I have self-reviewed my code
* [x] No new warnings are generated

*Test coverage*
* [ ] New code is covered by passing tests

lindsve committed:
Add OpenTelemetry library and expose metrics for Witsml requests

eliasbruvik commented:
If this is only supported with .NET 8, we will also need to update the version in the dockerfile for our internal server setup. We should also give a heads up and enough time for anyone else using Witsml Explorer that we will no longer support version 7. At least unless it's possible to only enable this when using .NET 8.

lindsve commented:
> If this is only supported with .NET 8, we will also need to update the version in the dockerfile for our internal server setup. We should also give a heads up and enough time for anyone else using Witsml Explorer that we will no longer support version 7. At least unless it's possible to only enable this when using .NET 8.

Good call. I'm not sure if I'm able to fully test this in a .NET 7 environment since I have .NET 8 sdk installed (the Witsml library seems to support multiple versions, currently .NET 6 and 7 although I added .NET 8 in this PR). I'll see if I can create a dummy .NET 7 project and reference a locally built witsml package and see how it goes.

That being said, .NET 8 is the latest LTS release and should be favored over .NET 7. I've also added a PR for updating to .NET 8 here: https://github.com/equinor/witsml-explorer/pull/2293

lindsve committed:
Fix build issue

lindsve committed:
Fix formatting

lindsve commented:
This PR also fixes an issue with duplicate references in Witsml.csproj that was not discovered in the previous PR (#2293). Discovered in the action that builds and publishes the package after merge.

eliasbruvik commented on Src/Witsml/Metrics/WitsmlMethod.cs:
```diff
@@ -0,0 +1,10 @@
+namespace Witsml.Metrics;
+
+public enum WitsmlMethod
```
If we want to measure all the store functions we are currently using, we also need to add GetVersion. It's not used a lot, so if you don't think it's necessary that's fine for me. If you add it, it's the `TestConnectionAsync` method in `WitsmlClient.cs` that uses it.

lindsve replied:
I need to check again, but I seem to recall that there was a reason for leaving that one out. Can't remember why though. Will check 👍

lindsve replied:
The reason was that `Result` in `WMLS_GetVersionResponse` for some reason is generated as string instead of short as other response types. Refactored to use string values for the result codes to include GetVersion in the measurements

eliasbruvik commented on Src/Witsml/Metrics/WitsmlMetrics.cs:
```diff
@@ -0,0 +1,64 @@
+
+        _activeRequests.Add(1, tagList);
+        var timer = Stopwatch.StartNew();
+        TResponseType response = await wmlsTask;
```
Just to verify: Are there any edge cases where wmlsTask would completely fail instead of returning a proper response? If it's possible, the _activeRequests wouldn't get reset.

I tested with malformatted requests and always got a response, but we might get problems if the server is unreachable. Sounds like a good idea to handle exceptions anyways.

lindsve replied:
Good call. I don't think that I have ever seen a missing response, but I'll add some exception handling to make sure that we don't end up out of sync anyway

lindsve committed:
Change type for ResultCode to string so that GetVersion can be included as well

lindsve committed:
Use try/finally to make sure that counter/timer are stopped if any exceptions are thrown in the fetch methods

eliasbruvik commented on Src/Witsml/WitsmlClient.cs:
```diff
@@ -432,11 +432,15 @@ public async Task<string> DeleteFromStoreAsync(string query, OptionsIn optionsIn
+            WMLS_GetVersionResponse response =
+                await _witsmlMetrics.MeasureQuery(
+                    _serverUrl,
+                    WitsmlMethod.GetCap,
```
A little oversight - you used GetCap instead of GetVersion (and forgot to add GetVersion to `WitsmlMethod.cs`)

lindsve committed:
Fix copy/paste error when adding measurement of TestConnection

eliasbruvik reviewed (approved):
- [x] Read issue and PR
- [x] Pulled branch and manually tested
- [x] Verify that PR resolves issue
- [x] Reviewed the code
